crates/ruff/src/rules/perflint/rules/slow_dict_creation.rs

```diff
@@ -0,0 +1,98 @@
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::prelude::{ast, Expr, Stmt};
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for cases where a new dictionary is made as a copy of an existing one by setting values
+/// in a for-loop
+///
+/// ## Why is this bad?
+/// It is more performant to use a dict comprehension to construct dictionaries
+///
+/// ## Example
+/// ```python
+/// pairs = (("a", 1), ("b", 2))
+/// result = {}
+/// for x, y in pairs:
+///     if y % 2:
+///         result[x] = y
+/// ```
+///
+/// Use instead:
+/// ```python
+/// pairs = (("a", 1), ("b", 2))
+/// result = {x: y for x, y in pairs if y % 2}
+/// ```
+#[violation]
+pub struct SlowDictCreation;
+
+impl Violation for SlowDictCreation {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Use a dict comprehension to create a (filtered) copy of a dict")
+    }
+}
+
+/// PERF404
+pub(crate) fn slow_dict_creation(checker: &mut Checker, target: &Expr, body: &[Stmt]) {
+    if body.len() != 1 {
+        return;
+    }
```

Comment:
Nit. I believe you can merge the `if body.len() != 1` and `let stmt = &body[0]` by doing

```suggestion
	let [stmt] = body else {
		return;
	}
```